### ISSUE TITLE
CS0465 (Introducing Finalize method) on interfaces

### ISSUE THREAD

**Issue by miloush** [dotnet/roslyn#20241]
**Version Used**: 15.3 Preview 2

**Steps to Reproduce**:
```C#
interface IBla
{
    void Finalize();
}
```
**Actual Behavior**:  CS0465: Introducing a 'Finalize' method can interfere with destructor invocation. Did you intend to declare a destructor?

**Expected Behavior**: Does this warning make sense on interfaces? I definitely can't declare a destructor and I am not exactly sure how the interface declaration could interfere with the destructor invocation either.

**Comment by sharwell** [comment]
:memo: The internal name for a destructor is `Finalize`. C# uses the `~TypeName()` syntax to define one but the compiler emits the method as one that overrides `Object.Finalize`. Someone from the compiler team will be able to give you more specifics about the intent of this particular warning.

**Comment by jaredpar** [comment]
The issue here is more about the problems it will create when the interface is used.  Consider the following:

``` c#
class C : IBla {
  ~C() { } 
  public void Finalize() { } 
}
```

This code won't compile because `~C` maps down to `Finalize` (as @sharwell pointed out). Hence this creates a conflict. So while the interface doesn't directly have a problem with `Finalize`, it's indirectly creating one for anyone who implements it.

**Comment by miloush** [comment]
OK so the 'Did you intend to declare a destructor?' part was not intended for interface authors.

```C#
class C {
  ~C() { } 
  public void Finalize() { } 
}
```
would fail compilation regardless of whether it implements the interface or not.

So theoretically all objects would implement such interface, except for that the Finalize method is not public, so it cannot implement interface methods. And directly declaring Finalize method is not allowed.

Basically the interface can be implemented only explicitly. 
Thanks!
